src/flux/scores.coffee

```diff
@@ -10,105 +11,111 @@ REJECTING = 'rejecting'
+  taskId = parseInt(taskId, 10)
+  for courseId, period of data
+    for period, periodIndex in data[courseId]
+      for student, studentIndex in period.students
+        for task in student.data
+          if task.id is taskId
+            return {task, courseId, period, periodIndex, studentIndex}
 
-ScoresConfig = {
-
-  _asyncStatus: {}
-
-
-  getTaskById: (taskId, courseId) ->
-    students = allStudents @_get(courseId)
-    data = _.flatten(_.pluck(students, 'data'))
-    task = _.findWhere(data, {id: taskId})
-    task
-
-  updateAverages: (task, courseId, period_id, columnIndex, currentValue, acceptValue) ->
-    scores = @_get(courseId)
-    period = _.findWhere(scores, {period_id})
-    periodTasks = _.flatten(_.pluck(period.students, 'data'))
-
-    numStudents = period.students.length
-    currentValue = currentValue / 100
-    acceptValue = acceptValue / 100
 
+adjustTaskAverages = (data, taskInfo) ->
+  {task} = taskInfo
+  oldScore = task.score
+  task.score = Math.round((
+    (task.correct_on_time_exercise_count + task.correct_accepted_late_exercise_count ) /
+      task.exercise_count
+  ) * 100 ) / 100
 
-    currentAssignmentAverage = period?.data_headings[columnIndex]?.average_score
+  course = data[taskInfo.courseId][0]
 
-    assignmentAverage =
-      (currentAssignmentAverage - (currentValue / numStudents)) +
-      (acceptValue / numStudents)
+  studentCount = course.students.length
+  heading = course.data_headings[taskInfo.periodIndex]
+  heading.average_score =
+    ( heading.average_score - ( oldScore / studentCount ) ) +
+      ( task.score / studentCount )
 
-    period?.data_headings[columnIndex]?.average_score = assignmentAverage
+  taskCount = 0
+  for student in course.students
+    taskCount += 1 for studentTask in student.data when studentTask.is_included_in_averages
 
+  course.overall_average_score =
+    (course.overall_average_score - ( oldScore / taskCount ) ) +
+      ( task.score / taskCount )
 
-    # number of included homeworks tasks in period
-    numStudentTasks = _.filter(periodTasks, {is_included_in_averages: true}).length
-
-    currentPeriodAverage = period?.overall_average_score
-
-    periodAverage =
-      (currentPeriodAverage - (currentValue / numStudentTasks)) +
-      (acceptValue / numStudentTasks)
-
-    period?.overall_average_score = periodAverage
-
-
-
-    students = allStudents @_get(courseId)
-    taskId = parseInt(task.id)
-
-    taskStudent =
-      _.find students, (student) ->
-        taskIds = _.pluck student.data, 'id'
-        _.indexOf(taskIds, taskId) > -1
 
+ScoresConfig = {
 
-    currentStudentAverage = taskStudent?.average_score
+  _asyncStatus: {}
 
-    # number of included homework tasks for student
-    numTasksStudent = _.filter(taskStudent.data, {is_included_in_averages: true}).length
+  hasLateWork: (task) ->
+    task.completed_step_count > task.completed_on_time_step_count +
+      task.completed_accepted_late_step_count
```

Comment:
@rvnewell has said that reading progress and homework progress should just based on exercise counts, not generic step counts.  So I think this function should change to reference exercise counts.  Even though I provide the generic counts, I don't think they should be used anywhere in the scores report.